File: String/Python/_0383_RansomNote.py
class Solution:
    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(1)
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        if len(magazine) < len(ransomNote):
            return False
        for char in ransomNote:
            if magazine.find(char) < 0:
                return False
            else:
                magazine = magazine.replace(char, '', 1)
        return True


ins = Solution()

print(ins.canConstruct("a", "b"))
print(ins.canConstruct("aa", "ab"))
print(ins.canConstruct("aa", "aab"))

File: String/TypeScript/_0383_RansomNote.ts
/**
 * Iteration
 * Time Complexity: BigO(N)
 * Space Complexity: BigO(N)
 */
function canConstruct(ransomNote: string, magazine: string): boolean {
  let store = magazine.split("");
  for (let letter of ransomNote) {
    if (store.includes(letter)) {
      let index = store.indexOf(letter);
      store.splice(index, 1);
    } else return false;
  }
  return true;
}

console.log(canConstruct("a", "b"));
console.log(canConstruct("aa", "ab"));
console.log(canConstruct("aa", "aab"));
